File: engine/tests/test_golden_parity.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference types="node" />
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const financeCalculator_1 = require("../financeCalculator");
const specDir = path.join(__dirname, '../../finance-spec');
const testsFile = path.join(specDir, 'golden-tests.json');
const policyFile = path.join(specDir, 'calculation-policy.json');
const goldenTests = JSON.parse(fs.readFileSync(testsFile, 'utf8'));
const policy = JSON.parse(fs.readFileSync(policyFile, 'utf8'));
function calculateGoldenTestWrapper(inputs, policy) {
    const scenario = inputs.scenario;
    const units = inputs.monthly_units_sold;
    const var_cost = inputs.variable_cost_per_unit;
    const fixed_cost = inputs.monthly_fixed_cost || 0;
    const transport = inputs.monthly_transport_cost || 0;
    const labour = inputs.monthly_labour_cost || 0;
    const rent = inputs.monthly_rent || 0;
    const other_fixed = inputs.monthly_other_fixed_cost || 0;
    const sim_outputs = {};
    if (scenario === "DEMAND_DROP_20" && units !== undefined) {
        inputs.monthly_units_sold = Math.floor(units * 0.8);
        sim_outputs.simulated_monthly_units_sold = inputs.monthly_units_sold;
    }
    else if (scenario === "RAW_MATERIAL_UP_20" && var_cost !== undefined) {
        inputs.variable_cost_per_unit = var_cost * 1.20;
        sim_outputs.simulated_variable_cost_per_unit = inputs.variable_cost_per_unit;
        if (units !== undefined) {
            sim_outputs.simulated_monthly_units_sold = units;
        }
    }
    else if (scenario === "TRANSPORT_COST_SPIKE_50") {
        const new_transport = transport * 1.50;
        inputs.monthly_fixed_cost = labour + rent + new_transport + other_fixed;
        sim_outputs.simulated_monthly_transport_cost = new_transport;
        if (units !== undefined) {
            sim_outputs.simulated_monthly_units_sold = units;
        }
    }
    const res = (0, financeCalculator_1.calculateFinancialAssessment)(inputs, policy);
    if (scenario) {
        for (const k of ["monthly_revenue", "monthly_variable_cost", "monthly_fixed_cost", "monthly_operating_surplus", "business_cash_available_for_debt_service"]) {
            if (res[k] !== undefined) {
                res[`simulated_${k}`] = res[k];
            }
        }
    }
    return { ...res, ...sim_outputs };
}
let passed = 0;
let failed = 0;
for (const testCase of goldenTests) {
    const result = calculateGoldenTestWrapper(testCase.inputs, policy);
    let testPassed = true;
    for (const [key, expectedValue] of Object.entries(testCase.expected)) {
        const resVal = result[key];
        if (expectedValue === null) {
            if (resVal !== null && resVal !== undefined) {
                console.error(`[FAIL] ${testCase.name}: Mismatch for ${key}: expected null, got ${resVal}`);
                testPassed = false;
            }
        }
        else if (typeof expectedValue === 'string') {
            if (resVal !== expectedValue) {
                console.error(`[FAIL] ${testCase.name}: Mismatch for ${key}: expected ${expectedValue}, got ${resVal}`);
                testPassed = false;
            }
        }
        else if (Array.isArray(expectedValue)) {
            if (String(resVal) !== String(expectedValue)) {
                console.error(`[FAIL] ${testCase.name}: Mismatch for ${key}: expected ${expectedValue}, got ${resVal}`);
                testPassed = false;
            }
        }
        else {
            if (Number(expectedValue) !== Number(resVal)) {
                console.error(`[FAIL] ${testCase.name}: Mismatch for ${key}: expected ${expectedValue}, got ${resVal}`);
                testPassed = false;
            }
        }
    }
    if (testPassed) {
        console.log(`[PASS] ${testCase.name}`);
        passed++;
    }
    else {
        failed++;
    }
}
console.log(`\nTests passed: ${passed}, failed: ${failed}`);
if (failed > 0) {
    process.exit(1);
}

File: backend/tests/test_golden_parity.py
import json
import os
import pytest
from decimal import Decimal, ROUND_HALF_UP, getcontext
from pathlib import Path

# Adjust path to find the spec file relative to this test
SPEC_DIR = Path(__file__).parent.parent.parent / "finance-spec"
GOLDEN_TESTS_FILE = SPEC_DIR / "golden-tests.json"
POLICY_FILE = SPEC_DIR / "calculation-policy.json"

# We will import our new pure calculator here
from app.services.finance_calculator import calculate_financial_assessment

def load_tests():
    with open(GOLDEN_TESTS_FILE, "r") as f:
        return json.load(f)

def load_policy():
    with open(POLICY_FILE, "r") as f:
        return json.load(f)

GOLDEN_TESTS = load_tests()
POLICY = load_policy()

def calculate_golden_test_wrapper(inputs, policy):
    # Scenario logic moved to test harness
    scenario = inputs.get("scenario")
    
    units = inputs.get("monthly_units_sold")
    var_cost = inputs.get("variable_cost_per_unit")
    fixed_cost = inputs.get("monthly_fixed_cost", 0)
    transport = inputs.get("monthly_transport_cost", 0)
    labour = inputs.get("monthly_labour_cost", 0)
    rent = inputs.get("monthly_rent", 0)
    other_fixed = inputs.get("monthly_other_fixed_cost", 0)
    
    sim_outputs = {}
    
    if scenario == "DEMAND_DROP_20" and units is not None:
        inputs["monthly_units_sold"] = int(units * 0.8)
        sim_outputs["simulated_monthly_units_sold"] = inputs["monthly_units_sold"]
        
    elif scenario == "RAW_MATERIAL_UP_20" and var_cost is not None:
        inputs["variable_cost_per_unit"] = float(Decimal(str(var_cost)) * Decimal("1.20"))
        sim_outputs["simulated_variable_cost_per_unit"] = inputs["variable_cost_per_unit"]
        if units is not None:
            sim_outputs["simulated_monthly_units_sold"] = units
            
    elif scenario == "TRANSPORT_COST_SPIKE_50":
        new_transport = float(Decimal(str(transport)) * Decimal("1.50"))
        inputs["monthly_fixed_cost"] = labour + rent + new_transport + other_fixed
        sim_outputs["simulated_monthly_transport_cost"] = new_transport
        if units is not None:
            sim_outputs["simulated_monthly_units_sold"] = units

    res = calculate_financial_assessment(inputs, policy)
    
    if scenario:
        for k in ["monthly_revenue", "monthly_variable_cost", "monthly_fixed_cost", "monthly_operating_surplus"]:
            if k in res:
                res[f"simulated_{k}"] = res[k]
                
    res.update(sim_outputs)
    return res

@pytest.mark.parametrize("test_case", GOLDEN_TESTS, ids=lambda tc: tc["name"])
def test_golden_parity(test_case):
    inputs = test_case["inputs"]
    expected = test_case["expected"]
    
    # Run the deterministic python engine
    result = calculate_golden_test_wrapper(inputs, POLICY)
    
    # Assert all expected fields match the result exactly
    for key, expected_value in expected.items():
        if expected_value is None:
            assert result.get(key) is None, f"Expected None for {key}, got {result.get(key)}"
        elif isinstance(expected_value, str):
            assert result.get(key) == expected_value, f"Mismatch for {key}: expected {expected_value}, got {result.get(key)}"
        elif isinstance(expected_value, list):
            assert result.get(key) == expected_value, f"Mismatch for {key}: expected {expected_value}, got {result.get(key)}"
        else:
            # Numeric comparison
            assert key in result, f"Missing key in result: {key}"
            res_val = result[key]
            if res_val is None:
                assert expected_value is None, f"Mismatch for {key}: expected {expected_value}, got None"
            else:
                expected_dec = Decimal(str(expected_value))
                res_dec = Decimal(str(res_val))
                assert expected_dec == res_dec, f"Mismatch for {key}: expected {expected_dec}, got {res_dec}"

